### PR TITLE
add introduction of dm's safe-mode

### PR DESCRIPTION
**Pull request by sunzhaoyang**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v6.0 (TiDB 6.0 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [x] Might cause conflicts after applied to another branch

### CODE REVIEW

File: dm/dm-safe-mode.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+---
+title: DM safe mode 介绍
+summary: 介绍 DM safe mode 作用和原理
+---
+
+# DM safe mode 介绍
+
+安全模式（safe mode）是 DM 在进行增量同步时候的一种运行模式，在安全模式中，DM 增量同步组件在执行 binlog event 时，将把所有 INSERT 和 UPDATE 操作强制进行改写后再在下游执行。
```
Comment (lance6716) [inline dm/dm-safe-mode.md:8]:
```suggestion
安全模式（safe mode）是 DM 在进行增量同步时候的一种运行模式，在安全模式中，DM 增量同步组件在同步 binlog event 时，将把所有 INSERT 和 UPDATE 操作强制进行改写后再在下游执行。
```

File: dm/dm-safe-mode.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+安全模式的目的是为了解决当增量同步过程中，同一条 binlog event 能够在下游被重复执行且保证幂等性，从而确保增量同步能够“安全”进行。
```
Comment (lance6716) [inline dm/dm-safe-mode.md:10]:
```suggestion
安全模式的目的是在增量同步过程中，同一条 binlog event 能够在下游被重复同步且保证幂等性，从而确保增量同步能够“安全”进行。 
```

File: dm/dm-safe-mode.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+安全模式的目的是为了解决当增量同步过程中，同一条 binlog event 能够在下游被重复执行且保证幂等性，从而确保增量同步能够“安全”进行。 
+
+最早引入安全模式是为了解决从 DM checkpoint 恢复后，某些 binlog 事件可能被重复执行而产生的问题。在进行增量同步过程中， DML 执行 和写 checkpoint 操作并不是同步的，并且写 checkpoint 操作和写下游数据也并不能保证原子性，当 DM 因为某些原因异常退出时，checkpoint 可能只记录到退出时刻之前的一个恢复点，因此当同步任务重启，从之前保存的 checkpoint 重新开始增量数据同步时，有一些 checkpoint 之后的数据可能已经在异常退出前被处理过了，这里会被重复处理。如果重复执行插入操作，会导致主键 / 唯一索引冲突导致同步中断；如果重复执行更新操作，会导致不能根据筛选条件找到之前对应的更新记录。安全模式通过改写 SQL 的方法能很好解决上述问题。
+
+## safe mode 原理
+
+安全模式原理是通过 SQL 语句改写来完成。具体来说，在安全模式下，所有的 INSERT 语句会被改写成 REPLACE 语句，而 UPDATE 语句会被分析，得到该语句涉及的行的主键 / 唯一索引的值，然后改写成 DELETE + REPLACE 语句 ：先根据主键 / 唯一索引的定位删除对应的行，然后使用REPLACE语句插入一条最新值的行记录。
+
+REPLACE 操作是 MySQL 特有的数据插入语法，它在插入数据时，如果发现新插入的数据和现有数据存在主键冲突，它会把原先的具有该主键的记录先删除，然后再执行插入记录操作。相当于“强制插入”的操作。具体请参考[MySQL 官方文档的 REPLACE 语句相关介绍](https://dev.mysql.com/doc/refman/8.0/en/replace.html) 。
```
Comment (lance6716) [inline dm/dm-safe-mode.md:18]:
```suggestion
REPLACE 操作是 MySQL 特有的数据插入语法，它在插入数据时，如果发现新插入的数据和现有数据存在主键或唯一约束冲突，它会把所有冲突的记录先删除，然后再执行插入记录操作。相当于“强制插入”的操作。具体请参考[MySQL 官方文档的 REPLACE 语句相关介绍](https://dev.mysql.com/doc/refman/8.0/en/replace.html) 。
```

File: dm/dm-safe-mode.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+从 checkpoint 进行恢复增量同步任务时，DM 会根据下面的判断逻辑来开启安全模式：
+
+- 如果 checkpoint 信息里面额外记录了 safemode_exit_point ，说明之前是异常暂停。如果恢复时发现待恢复的 checkpoint 的 binlog 位置点 < safemode_exit_point，那么说明从这个 checkpoint 到 safemode_exit_point 之间的 binlog 可能已经在下游被执行了，在这里要被重复执行。所以这段 binlog 范围内需要开启安全模式。直到当前的binlog 位置点超过了 safemode_exit_point，DM 才会考虑关闭安全模式。
+
+- 如果 checkpoint 信息里面没有 safemode_exit_point 信息，一种可能是说明这是一个全新的任务或者之前该任务是正常暂停。但是，还有一种可能是之前异常暂停时，记录 safemode_exit_point 的时候失败了，甚至可能是之前 DM 进程异常退出了。这种情况下，DM 并不知道哪些 checkpoint 之后的 binlog 已经被执行过了。所以，这种情况下，为了保险起见，只要没有在 checkpoint 找到 safemode_exit_point 信息，DM 就会在前2个 checkpoint 间隔中都开启安全模式，确保这段时间的重复执行 binlog 都没有问题。默认的 checkpoint 间隔是30秒，也就是说，一个正常的增量同步任务开始时，前 2 * 30 = 60秒会自动强制开启安全模式。通过设置任务配置中 syncer的配置 “checkpoint-flush-interval” 来调整 checkpoint 的间隔，从而影响增量同步任务开始时安全模式的持续时间（一般情况下不建议调整，如有需要应优先使用后文所述的`手动开启`）。
+
+另外，当正在进行一组分片表的 DDL 操作时，如果分片表 DDL 的处理模式是“悲观协调”模式，并且当时上游没有完成所有涉及分片表的 DDL 操作，此时增量同步会暂时设置为安全模式。这样设计的目的是能够让一组分片表在涉及主键 / 唯一索引修改的 DDL 操作时，在下游能够在新的主键 / 唯一索引生效前，避免旧的主键 / 唯一索引冲突。比如，上游分片表当中，有一些表完成了唯一索引的修改后，由于下游需要等待所有上游分区表的唯一索引全部生效才会更新唯一索引，此时下游还在使用老的唯一索引。此时有一定的概率出现在上游分区表中，新的唯一索引不存在冲突，但是在老的唯一索引约束下会存在冲突的修改，这些修改到了下游之后，就会存在下游的唯一索引的冲突。所以此时开启安全模式能够解决这个问题。
```
Comment (lance6716) [inline dm/dm-safe-mode.md:49]: 原理错误吧，”有一些表完成了唯一索引的修改后“，这些表的 DML 会 block。 @sunzhaoyang 可以问问原作者的理解

File: dm/dm-safe-mode.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+## 如何开启 safe mode
+
+### 自动开启
+
+当 DM 从 checkpoint 恢复增量同步任务(例如 worker 重启，网络中断重连等)时，都会自动开启一段时间的安全模式，原因会在下面详细分析。在 checkpoint 中，有一个叫做 safemode_exit_point 的信息和安全模式的开启逻辑息息相关。当上一个增量同步任务异常暂停时，DM 会先尝试将内存中所有的 DML 全部同步到下游，然后 DM 记录当前内存中从上游拉取到的最新的 binlog 位置点，记作 safemode_exit_point，把这个额外信息保存在异常暂停之前最后一个 checkpoint 当中。
+
+从 checkpoint 进行恢复增量同步任务时，DM 会根据下面的判断逻辑来开启安全模式：
+
+- 如果 checkpoint 信息里面额外记录了 safemode_exit_point ，说明之前是异常暂停。如果恢复时发现待恢复的 checkpoint 的 binlog 位置点 < safemode_exit_point，那么说明从这个 checkpoint 到 safemode_exit_point 之间的 binlog 可能已经在下游被执行了，在这里要被重复执行。所以这段 binlog 范围内需要开启安全模式。直到当前的binlog 位置点超过了 safemode_exit_point，DM 才会考虑关闭安全模式。
```
Comment (Ehco1996) [inline dm/dm-safe-mode.md:45]:
```suggestion
- 如果 checkpoint 信息里面额外记录了 safemode_exit_point ，说明之前是异常暂停。如果恢复时发现待恢复的 checkpoint 的 binlog 位置点 < safemode_exit_point，那么说明从这个 checkpoint 到 safemode_exit_point 之间的 binlog 可能已经在下游被执行了，在这里要被重复执行。所以这段 binlog 范围内需要开启安全模式。直到当前的binlog 位置点超过了 safemode_exit_point，DM 才会关闭安全模式。
```

File: dm/dm-safe-mode.md (hunk 4 above)
Comment (lance6716) [reply dm/dm-safe-mode.md:49]: @dsdashun

Comment (dsdashun) [reply dm/dm-safe-mode.md:49]: 这块可能是有一些问题，我会确认好之后看是不是要把这一段删掉

File: dm/dm-safe-mode.md (hunk 5 above)
Comment (dsdashun) [reply dm/dm-safe-mode.md:45]: 这里写成“考虑关闭”是因为如果在配置设置了safe-mode = true，安全模式是关不掉的

File: dm/dm-safe-mode.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+- 如果 checkpoint 信息里面没有 safemode_exit_point 信息，一种可能是说明这是一个全新的任务或者之前该任务是正常暂停。但是，还有一种可能是之前异常暂停时，记录 safemode_exit_point 的时候失败了，甚至可能是之前 DM 进程异常退出了。这种情况下，DM 并不知道哪些 checkpoint 之后的 binlog 已经被执行过了。所以，这种情况下，为了保险起见，只要没有在 checkpoint 找到 safemode_exit_point 信息，DM 就会在前2个 checkpoint 间隔中都开启安全模式，确保这段时间的重复执行 binlog 都没有问题。默认的 checkpoint 间隔是30秒，也就是说，一个正常的增量同步任务开始时，前 2 * 30 = 60秒会自动强制开启安全模式。通过设置任务配置中 syncer的配置 “checkpoint-flush-interval” 来调整 checkpoint 的间隔，从而影响增量同步任务开始时安全模式的持续时间（一般情况下不建议调整，如有需要应优先使用后文所述的`手动开启`）。
+
+另外，当正在进行一组分片表的 DDL 操作时，如果分片表 DDL 的处理模式是“悲观协调”模式，并且当时上游没有完成所有涉及分片表的 DDL 操作，此时增量同步会暂时设置为安全模式。这样设计的目的是能够让一组分片表在涉及主键 / 唯一索引修改的 DDL 操作时，在下游能够在新的主键 / 唯一索引生效前，避免旧的主键 / 唯一索引冲突。比如，上游分片表当中，有一些表完成了唯一索引的修改后，由于下游需要等待所有上游分区表的唯一索引全部生效才会更新唯一索引，此时下游还在使用老的唯一索引。此时有一定的概率出现在上游分区表中，新的唯一索引不存在冲突，但是在老的唯一索引约束下会存在冲突的修改，这些修改到了下游之后，就会存在下游的唯一索引的冲突。所以此时开启安全模式能够解决这个问题。
```
Comment (sunzhaoyang) [inline dm/dm-safe-mode.md:49]:
```suggestion
```

File: dm/dm-safe-mode.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    syncer-config-name: "global"            # syncers 配置的名称
+```
+
+## 注意事项
+
+使用安全模式主要解决的是由于之前同步任务的异常退出而导致的恢复后 binlog 重复执行的问题。有一些用户为了安全起见，可能会通过配置全程开启安全模式。但是，如果全程开启安全模式，有一些问题需要注意：
+
+- 安全模式对于增量同步有额外开销。这是因为频繁的 DELETE + REPLACE 操作会带来主键 / 唯一索引的频繁变化，带来了比单纯 UPDATE 语句更大的性能开销。
+- 由于安全模式会强制替换主键相同的记录，带来的问题就是之前的记录数据丢失。如果由于上游分片表合并导入下游的配置不当，可能导致在上游合并导入到下游数据库的时候，出现大量的主键 / 唯一索引冲突。如果此时全程开启安全模式，会导致下游大量的数据丢失，并且可能无法从任务中看到有任何的异常，导致数据的大量不一致。
+
```
Comment (dsdashun) [inline dm/dm-safe-mode.md:74]:
```suggestion
- 安全模式依赖于通过主键 / 唯一索引来判断冲突。如果下游 DB 对应的表没有主键 / 唯一索引，会导致 REPLACE 语句起不到替换插入的目的，这样即使开启了安全模式，对于INSERT语句的改写成REPLACE并执行后，依旧会向下游插入重复记录。

```

File: dm/dm-safe-mode.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+安全模式原理是通过 SQL 语句改写来完成。具体来说，在安全模式下，所有的 INSERT 语句会被改写成 REPLACE 语句，而 UPDATE 语句会被分析，得到该语句涉及的行的主键 / 唯一索引的值，然后改写成 DELETE + REPLACE 语句 ：先根据主键 / 唯一索引的定位删除对应的行，然后使用REPLACE语句插入一条最新值的行记录。
+
+REPLACE 操作是 MySQL 特有的数据插入语法，它在插入数据时，如果发现新插入的数据和现有数据存在主键或唯一约束冲突，它会把所有冲突的记录先删除，然后再执行插入记录操作。相当于“强制插入”的操作。具体请参考[MySQL 官方文档的 REPLACE 语句相关介绍](https://dev.mysql.com/doc/refman/8.0/en/replace.html) 。
+
+比如，如果假设一张表`dummydb.dummytbl`，主键是`id`，并且下面的 SQL 被重复执行：
+
+```
+INSERT INTO dummydb.dummytbl (id, int_value, str_value) VALUES (123, 999, 'abc');
+UPDATE dummydb.dummytbl SET int_value = 888999 WHERE int_value = 999；   # 假设没有其他 int_value = 999的数据
```
Comment (dsdashun) [inline dm/dm-safe-mode.md:24]:
```suggestion
UPDATE dummydb.dummytbl SET int_value = 888999 WHERE int_value = 999；   # 假设没有其他 int_value = 999的数据
UPDATE dummydb.dummytbl SET id = 999 WHERE id = 888；    # 更新主键操作
```

File: dm/dm-safe-mode.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+```
+
+经过安全模式的 SQL 改写，再次在下游执行时会变成执行下面的 SQL 语句：
+
+```
+REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (123, 999, 'abc');
+DELETE FROM dummydb.dummytbl WHERE id = 123;
+REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (123, 888999, 'abc')；
```
Comment (dsdashun) [inline dm/dm-safe-mode.md:32]:
```suggestion
REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (123, 888999, 'abc')；
DELETE FROM dummydb.dummytbl WHERE id = 888;
REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (999, 888888, 'abc888');    # 更新主键的改写解释了为什么 UPDATE 要 替换为 DELETE + REPLACE, 而不是 DELETE + INSERT ：如果这里使用INSERT，那么id = 999 的记录在已经存在的情况下重复插入时会报错主键冲突，而REPLACE则会替换之前已经插入的记录
```

**Review by dsdashun — APPROVED** [review]
LGTM

File: dm/dm-safe-mode.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+summary: 介绍 DM safe mode 作用和原理
+---
+
+# DM safe mode 介绍
+
+安全模式（safe mode）是 DM 在进行增量同步时候的一种运行模式，在安全模式中，DM 增量同步组件在同步 binlog event 时，将把所有 INSERT 和 UPDATE 操作强制进行改写后再在下游执行。
+
+安全模式的目的是在增量同步过程中，同一条 binlog event 能够在下游被重复同步且保证幂等性，从而确保增量同步能够“安全”进行。 
+
+最早引入安全模式是为了解决从 DM checkpoint 恢复后，某些 binlog 事件可能被重复执行而产生的问题。在进行增量同步过程中， DML 执行 和写 checkpoint 操作并不是同步的，并且写 checkpoint 操作和写下游数据也并不能保证原子性，当 DM 因为某些原因异常退出时，checkpoint 可能只记录到退出时刻之前的一个恢复点，因此当同步任务重启，从之前保存的 checkpoint 重新开始增量数据同步时，有一些 checkpoint 之后的数据可能已经在异常退出前被处理过了，这里会被重复处理。如果重复执行插入操作，会导致主键 / 唯一索引冲突导致同步中断；如果重复执行更新操作，会导致不能根据筛选条件找到之前对应的更新记录。安全模式通过改写 SQL 的方法能很好解决上述问题。
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:12]:
```suggestion
DM 从 checkpoint 恢复数据同步任务后，可能重复执行某些 binlog 事件而导致下述问题：

1. 在进行增量同步过程中，执行 DML 的操作和写 checkpoint 的操作并不是同步的；写 checkpoint 的操作和写下游数据的操作也并不能保证原子性。因此，**当 DM 异常退出时，checkpoint 可能只记录到退出时刻之前的一个恢复点**。
2. 当 DM 重启同步任务，并从 checkpoint 重新开始增量数据同步时，checkpoint 之后的部分数据可能已经在异常退出前被处理过了，从而**导致部分 SQL 语句重复执行**。
3. 如果重复执行 `INSERT` 操作，会导致主键或唯一索引冲突，引发同步中断；如果重复执行 `UPDATE` 操作，会导致不能根据筛选条件找到之前对应的更新记录。

在安全模式下，通过改写 SQL 语句，DM 可以解决上述问题。
```

File: dm/dm-safe-mode.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+## safe mode 原理
+
+安全模式原理是通过 SQL 语句改写来完成。具体来说，在安全模式下，所有的 INSERT 语句会被改写成 REPLACE 语句，而 UPDATE 语句会被分析，得到该语句涉及的行的主键 / 唯一索引的值，然后改写成 DELETE + REPLACE 语句 ：先根据主键 / 唯一索引的定位删除对应的行，然后使用REPLACE语句插入一条最新值的行记录。
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:16]:
```suggestion
安全模式通过 SQL 语句改写来保证 binlog event 的幂等性。具体来说，在安全模式下：

* `INSERT` 语句会被改写成 `REPLACE` 语句。
* `UPDATE` 语句会被分析，得到该语句涉及的行的主键或唯一索引的值，然后改写成 `DELETE` + `REPLACE` 语句 ：先根据主键或唯一索引的定位删除对应的行，然后使用 `REPLACE` 语句插入一条最新值的行记录。
```

File: dm/dm-safe-mode.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+REPLACE 操作是 MySQL 特有的数据插入语法，它在插入数据时，如果发现新插入的数据和现有数据存在主键或唯一约束冲突，它会把所有冲突的记录先删除，然后再执行插入记录操作。相当于“强制插入”的操作。具体请参考[MySQL 官方文档的 REPLACE 语句相关介绍](https://dev.mysql.com/doc/refman/8.0/en/replace.html) 。
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:18]:
```suggestion
`REPLACE` 操作是 MySQL 特有的数据插入语法。使用 `REPLACE` 语法插入数据时，如果新插入的数据和现有数据存在主键或唯一约束冲突，MySQL 会删除所有冲突的记录，然后再执行插入记录操作。相当于“强制插入”的操作。具体请参考 [MySQL 官方文档的 `REPLACE` 语句相关介绍](https://dev.mysql.com/doc/refman/8.0/en/replace.html) 。
```

File: dm/dm-safe-mode.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+比如，如果假设一张表`dummydb.dummytbl`，主键是`id`，并且下面的 SQL 被重复执行：
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:20]:
```suggestion
比如，假设一张表 `dummydb.dummytbl` 的主键是 `id`，在这张表中重复执行下面的 SQL 语句：
```

File: dm/dm-safe-mode.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+```
+INSERT INTO dummydb.dummytbl (id, int_value, str_value) VALUES (123, 999, 'abc');
+UPDATE dummydb.dummytbl SET int_value = 888999 WHERE int_value = 999；   # 假设没有其他 int_value = 999的数据
+UPDATE dummydb.dummytbl SET id = 999 WHERE id = 888；    # 更新主键操作
+```
+
+经过安全模式的 SQL 改写，再次在下游执行时会变成执行下面的 SQL 语句：
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:28]:
```suggestion
启用安全模式后，上述 SQL 语句再次在下游执行时，会被改写为下面的 SQL 语句：
```

File: dm/dm-safe-mode.md [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+```
+REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (123, 999, 'abc');
+DELETE FROM dummydb.dummytbl WHERE id = 123;
+REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (123, 888999, 'abc')；
+DELETE FROM dummydb.dummytbl WHERE id = 888;
+REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (999, 888888, 'abc888');    # 更新主键的改写解释了为什么 UPDATE 要 替换为 DELETE + REPLACE, 而不是 DELETE + INSERT ：如果这里使用INSERT，那么id = 999 的记录在已经存在的情况下重复插入时会报错主键冲突，而REPLACE则会替换之前已经插入的记录
+```
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:36]: 代码注释的可读性比较差，建议单独提出来写一行。

````suggestion
REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (999, 888888, 'abc888');    
```

上述语句中，`UPDATE` 被改写为 `DELETE` + `REPLACE`，而不是 `DELETE` + `INSERT`。如果这里使用 `INSERT`，那么 `id = 999` 的记录在已经存在的情况下重复插入时会报错主键冲突，而 `REPLACE` 则会替换之前已经插入的记录。
````

File: dm/dm-safe-mode.md [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+通过这样的语句改写，在进行重复的插入或更新操作时，都会把执行该操作后的行数据覆盖之前已经存在的行数据。这样保证了插入和更新操作的可重复执行。
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:38]:
```suggestion
通过这样的语句改写，在进行重复的插入或更新操作时，DM 都会使用执行该操作后的行数据来覆盖之前已经存在的行数据。这样保证了插入和更新操作的可重复执行。
```

**Review by ran-huang — COMMENTED** [review]
Reviewed contents above L40

File: dm/dm-safe-mode.md [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+
+## 注意事项
+
+使用安全模式主要解决的是由于之前同步任务的异常退出而导致的恢复后 binlog 重复执行的问题。有一些用户为了安全起见，可能会通过配置全程开启安全模式。但是，如果全程开启安全模式，有一些问题需要注意：
+
+- 安全模式对于增量同步有额外开销。这是因为频繁的 DELETE + REPLACE 操作会带来主键 / 唯一索引的频繁变化，带来了比单纯 UPDATE 语句更大的性能开销。
+- 由于安全模式会强制替换主键相同的记录，带来的问题就是之前的记录数据丢失。如果由于上游分片表合并导入下游的配置不当，可能导致在上游合并导入到下游数据库的时候，出现大量的主键 / 唯一索引冲突。如果此时全程开启安全模式，会导致下游大量的数据丢失，并且可能无法从任务中看到有任何的异常，导致数据的大量不一致。
+- 安全模式依赖于通过主键 / 唯一索引来判断冲突。如果下游 DB 对应的表没有主键 / 唯一索引，会导致 REPLACE 语句起不到替换插入的目的，这样即使开启了安全模式，对于INSERT语句的改写成REPLACE并执行后，依旧会向下游插入重复记录。
+
+因此，如果上游存在主键重复的数据，业务对于重复数据的丢失和性能损失可以接受，可以开启 safe mode 忽略数据重复错误。
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:90]:
```suggestion
因此，如果上游存在主键重复的数据，业务可以接受丢失重复数据和性能损失，你可以开启安全模式来忽略数据重复错误。
```

File: dm/dm-safe-mode.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+- 如果 checkpoint 信息里没有 `safemode_exit_point` 信息，则有两种情况：
+
+    1. 这是一个全新的任务，或该任务是正常暂停。
+    2. 该任务异常暂停时，记录 `safemode_exit_point` 失败；或是 DM 进程异常退出。
+    
+    第二种情况下，DM 并不知道哪些 checkpoint 之后的 binlog 已经被执行过。为了保险起见，只要没有在 checkpoint 找到 `safemode_exit_point` 信息，DM 就会在前 2 个 checkpoint 间隔中都开启安全模式，确保这段时间里的 binlog 被重复执行不会引发问题。默认的 checkpoint 间隔是 30 秒，也就是说，一个正常的增量同步任务开始时，前 2 * 30 = 60 秒会自动强制开启安全模式。
+    
+    你可以配置 syncer 的配置项 `checkpoint-flush-interval` 来调整 checkpoint 的间隔，从而影响增量同步任务开始时安全模式的持续时间。一般情况下不建议调整，如有需要，应优先使用[手动开启](#手动开启)。
+
+
```
Comment (sunzhaoyang) [inline dm/dm-safe-mode.md:70]:
```suggestion
```

File: dm/dm-safe-mode.md [hunk 19]
```diff
@@ -0,0 +1,96 @@
+---
+title: DM safe mode 介绍
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:2]:
```suggestion
title: DM 安全模式
```

File: dm/dm-safe-mode.md [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+summary: 介绍 DM safe mode 作用和原理
+---
+
+# DM safe mode 介绍
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:6]:
```suggestion
# DM 安全模式
```

File: dm/dm-safe-mode.md [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+
+安全模式（safe mode）是 DM 在进行增量同步时候的一种运行模式，在安全模式中，DM 增量同步组件在同步 binlog event 时，将把所有 INSERT 和 UPDATE 操作强制进行改写后再在下游执行。
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:8]:
```suggestion
安全模式 (safe mode) 是 DM 在进行增量同步时候的一种运行模式，在安全模式中，DM 增量同步组件在同步 binlog event 时，将把所有 `INSERT` 和 `UPDATE` 操作强制进行改写后再在下游执行。
```

File: dm/dm-safe-mode.md [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+
+DM 从 checkpoint 恢复数据同步任务后，可能重复执行某些 binlog 事件而导致下述问题：
+
+1. 在进行增量同步过程中，执行 DML 的操作和写 checkpoint 的操作并不是同步的；写 checkpoint 的操作和写下游数据的操作也并不能保证原子性。因此，**当 DM 异常退出时，checkpoint 可能只记录到退出时刻之前的一个恢复点**。
+2. 当 DM 重启同步任务，并从 checkpoint 重新开始增量数据同步时，checkpoint 之后的部分数据可能已经在异常退出前被处理过了，从而**导致部分 SQL 语句重复执行**。
+3. 如果重复执行 `INSERT` 操作，会导致主键或唯一索引冲突，引发同步中断；如果重复执行 `UPDATE` 操作，会导致不能根据筛选条件找到之前对应的更新记录。
+
+在安全模式下，通过改写 SQL 语句，DM 可以解决上述问题。
+
+## safe mode 原理
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:20]:
```suggestion
## 安全模式原理
```

File: dm/dm-safe-mode.md [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+安全模式通过 SQL 语句改写来保证 binlog event 的幂等性。具体来说，在安全模式下：
+
+* `INSERT` 语句会被改写成 `REPLACE` 语句。
+* `UPDATE` 语句会被分析，得到该语句涉及的行的主键或唯一索引的值，然后改写成 `DELETE` + `REPLACE` 语句 ：先根据主键或唯一索引的定位删除对应的行，然后使用 `REPLACE` 语句插入一条最新值的行记录。
+
+`REPLACE` 操作是 MySQL 特有的数据插入语法。使用 `REPLACE` 语法插入数据时，如果新插入的数据和现有数据存在主键或唯一约束冲突，MySQL 会删除所有冲突的记录，然后再执行插入记录操作。相当于“强制插入”的操作。具体请参考 [MySQL 官方文档的 `REPLACE` 语句相关介绍](https://dev.mysql.com/doc/refman/8.0/en/replace.html) 。
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:27]:
```suggestion
`REPLACE` 操作是 MySQL 特有的数据插入语法。使用 `REPLACE` 语法插入数据时，如果新插入的数据和现有数据存在主键或唯一约束冲突，MySQL 会删除所有冲突的记录，然后再执行插入记录操作，相当于“强制插入”的操作。具体请参考 [MySQL 官方文档的 `REPLACE` 语句相关介绍](https://dev.mysql.com/doc/refman/8.0/en/replace.html) 。
```

File: dm/dm-safe-mode.md [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+
+```
+REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (123, 999, 'abc');
+DELETE FROM dummydb.dummytbl WHERE id = 123;
+REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (123, 888999, 'abc')；
+DELETE FROM dummydb.dummytbl WHERE id = 888;
+REPLACE INTO dummydb.dummytbl (id, int_value, str_value) VALUES (999, 888888, 'abc888');    
+```
+
+上述语句中，`UPDATE` 被改写为 `DELETE` + `REPLACE`，而不是 `DELETE` + `INSERT`。如果这里使用 `INSERT`，那么 `id = 999` 的记录在已经存在的情况下重复插入时会报错主键冲突，而 `REPLACE` 则会替换之前已经插入的记录。
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:47]:
```suggestion
上述语句中，`UPDATE` 被改写为 `DELETE` + `REPLACE`，而不是 `DELETE` + `INSERT`。如果这里使用 `INSERT`，那么 `id = 999` 的记录在已经存在的情况下重复插入时，数据库会报错主键冲突。因此这里使用了 `REPLACE`，新的记录会替换之前已经插入的记录。
```

File: dm/dm-safe-mode.md [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+
+通过这样的语句改写，在进行重复的插入或更新操作时，DM 都会使用执行该操作后的行数据来覆盖之前已经存在的行数据。这样保证了插入和更新操作的可重复执行。
+
+## 如何开启 safe mode
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:51]:
```suggestion
## 开启安全模式
```

File: dm/dm-safe-mode.md [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+
+### 自动开启
+
+当 DM 从 checkpoint 恢复增量同步任务(例如 worker 重启，网络中断重连等)时，会自动开启一段时间的安全模式。开启安全模式的逻辑，与 checkpoint 中的 `safemode_exit_point` 信息相关。当一个增量同步任务异常暂停时，DM 会先尝试将内存中所有的 DML 全部同步到下游，然后记录当前内存中从上游拉取到的最新的 binlog 位置点，记作 `safemode_exit_point`，把 `safemode_exit_point` 保存在异常暂停之前最后一个 checkpoint 当中。
```
Comment (ran-huang) [inline dm/dm-safe-mode.md:55]:
```suggestion
当 DM 从 checkpoint 恢复增量同步任务（例如 worker 重启，网络中断重连等）时，会自动开启一段时间的安全模式。开启安全模式的逻辑，与 checkpoint 中的 `safemode_exit_point` 信息相关。当一个增量同步任务异常暂停时，DM 会先尝试将内存中所有的 DML 全部同步到下游，然后记录当前内存中从上游拉取到的最新的 binlog 位置点，记作 `safemode_exit_point`，把 `safemode_exit_point` 保存在异常暂停之前最后一个 checkpoint 当中。
```

**Review by ran-huang — APPROVED** [review]
Rest LGTM

**Pull request merged**